### PR TITLE
Bump actions/setup-go from 3 to 4

### PR DESCRIPTION
**Pull request by tushartathgur**
The new version of the action enables caching by default.

### CODE REVIEW

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -94,23 +77,9 @@ jobs:
     - name: Check-out code
       uses: actions/checkout@v3
     - name: Set up Go using version from go.mod
-      uses: actions/setup-go@v3
+      uses: actions/setup-go@v4
```
Comment (antoninbas) [inline .github/workflows/go.yml:80]: you need to set `cache-dependency-path` here, given that the relevant go.mod is not at the root of the repository

refer to https://github.com/actions/setup-go#caching-dependency-files-and-build-outputs

File: .github/workflows/golicense.yml [hunk 2]
```diff
@@ -35,7 +35,7 @@ jobs:
     steps:
     - uses: actions/checkout@v3
     - name: Set up Go using version from go.mod
-      uses: actions/setup-go@v3
+      uses: actions/setup-go@v4
```
Comment (antoninbas) [inline .github/workflows/golicense.yml:38]: you probably want to set `cache-dependency-path` here as well, to be safe
this time you would set it to `'**/go.sum'`

File: .github/workflows/process_release.yml [hunk 3]
```diff
@@ -10,9 +10,11 @@ jobs:
     steps:
     - uses: actions/checkout@v3
     - name: Set up Go using version from go.mod
-      uses: actions/setup-go@v3
+      uses: actions/setup-go@v4
       with:
         go-version-file: 'go.mod'
```
Comment (antoninbas) [inline .github/workflows/process_release.yml:15]: in this case, `cache-dependency-path` needs to be set to `'**/go.sum'`

File: .github/workflows/clair.yml [hunk 4]
```diff
@@ -11,7 +11,7 @@ jobs:
     runs-on: [ubuntu-latest]
     steps:
     - uses: actions/checkout@v3
```
Comment (antoninbas) [inline .github/workflows/clair.yml:13]: not related to this PR, but this whole workflow should probably be removed entirely as a follow-up
we moved to Trivy in antrea-io/antrea

File: .github/workflows/dependabot.yml [hunk 5]
```diff
@@ -37,7 +37,7 @@ jobs:
         ref: ${{ github.event.pull_request.head.sha }}
         token: ${{ secrets.ANTREA_BOT_WRITE_PAT }}
     - name: Set up Go using version from go.mod
-      uses: actions/setup-go@v3
+      uses: actions/setup-go@v4
```
Comment (antoninbas) [inline .github/workflows/dependabot.yml:40]: We probably need to set `cache-dependency-path` for this one

**Review by antoninbas — COMMENTED** [review]
Please review all the uses of the action, and determine if `cache-dependency-path` needs to be set. The rules are as follows:
1) If the only Go module being built is the one for which `go.mod` is at the root, then `cache-dependency-path` does not need to be set
2) If the only Go module being built is one for which the go.mod is not at the root (`snowflake/`) in our case, then `cache-dependency-path` needs to be set correctly (`snowflake/go.sum`).
3) If more than one Go module is being built (we only have 2 in our case), then use `cache-dependency-path: '**/go.sum'`. This is a safe way to cover all Go modules.

File: .github/workflows/go.yml [hunk 6]
```diff
@@ -181,9 +139,10 @@ jobs:
...
```
[the remark this answers is not in the document]
Comment (tushartathgur) [reply .github/workflows/go.yml:142]: little confused, why would we not require to setup go while make verify uses go?

[the remark this answers is not in the document]
Comment (antoninbas) [reply .github/workflows/go.yml:142]: ignore this, I forgot `make verify` was using Go and I was thinking of some Antrea CI job

File: .github/workflows/kind.yml [hunk 7]
```diff
@@ -116,9 +116,10 @@ jobs:
           sudo rm -rf "$AGENT_TOOLSDIRECTORY"
           df -h
       - uses: actions/checkout@v3
-      - uses: actions/setup-go@v3
+      - uses: actions/setup-go@v4
         with:
           go-version-file: 'go.mod'
+          cache-dependency-path: '**/go.sum'
```
[the remark this answers is not in the document]
Comment (tushartathgur) [reply .github/workflows/kind.yml:122]: do you think it is fine to keep this till we have not cleared clair-scan directory as per your initial comment, and could then be removed with the PR later?

[the remark this answers is not in the document]
Comment (antoninbas) [reply .github/workflows/kind.yml:122]: I don't mind, but I am not sure what the relationship is with the clair-scan stuff? The Kind jobs only build the top-level Go module, don't they?

**Review by antoninbas — APPROVED** [review]

File: .github/workflows/kind.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (tushartathgur) [reply .github/workflows/kind.yml:122]: that is correct!

**Review by antoninbas — APPROVED** [review]

**Pull request merged**